src/Data/Text/Internal/Reverse.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnliftedFFITypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+{-# OPTIONS_HADDOCK not-home #-}
+
+-- | Implements 'reverse', using efficient C routines by default.
+module Data.Text.Internal.Reverse (reverse) where
+
+#if !defined(PURE_HASKELL)
+import GHC.Exts as Exts
+import Control.Monad.ST.Unsafe (unsafeIOToST)
+import Foreign.C.Types (CSize(..))
+#else
+import Control.Monad (when)
+import Control.Monad.ST (ST)
+import Data.Text.Internal.Encoding.Utf8 (utf8LengthByLeader)
+#endif
+import Prelude hiding (reverse)
+import Data.Text.Internal (Text(..))
+import Control.Monad.ST (runST)
+import qualified Data.Text.Array as A
+
+-- | /O(n)/ Reverse the characters of a string.
+--
+-- Example:
+--
+-- $setup
+-- >>> T.reverse "desrever"
+-- "reversed"
+reverse ::
+#if defined(ASSERTS)
+  HasCallStack =>
+#endif
+  Text -> Text
+#if defined(PURE_HASKELL)
+reverse (Text ba off len) = runST $ do
+    dest <- A.new len
+    _ <- reversePoints ba off dest (len - 1)
+    result <- A.unsafeFreeze dest
+    pure $ Text result 0 len
+
+reversePoints
+    :: A.Array -- ^ Input array
+    -> Int -- ^ Input index
+    -> A.MArray s -- ^ Output array
+    -> Int -- ^ Output index
+    -> ST s ()
+reversePoints _ _ _ pOut | pOut < 0 = pure ()
+reversePoints src pIn dest pOut =
+    let pointLength = utf8LengthByLeader (A.unsafeIndex src pIn)
+        pOut' = pOut - pointLength + 1
+        -- Repeated unsafeWrite is faster than copyI
```

Comment:
Do you know which is faster with JS backend?

Reply:
I mean, unless we know JS backend behaviour, I'd strive for simpler Core, which should give GHC more opportunities to do backend-dependent optimizations.

Reply:
I'll double check. TBH, in many cases I'd rather have simpler code even if it's a little slower.

Reply:
@hsyl20 had suggested this change for backends *other* than JavaScript, by the way. tbh I'd rather use `copyI`. If `copyI` (a primop) is truly slower than repeated calls to `unsafeWrite` (another primop) maybe it's the primops that need to change.

Reply:
I've just used copyI because it seems like the right thing to do. I've squashed all commits as well.